### PR TITLE
Remove ActivitySourceAdapter

### PR DESCRIPTION
**Pull request by utpilla**
Currently, the instrumentation libraries depend on ActivitySourceAdapter to set the activity sampling to comply with the TracerProvider sampler. This PR demonstrates how we could implement instrumentation libraries without using ActivitySourceAdapter. I have updated AspNet, AspNetCore, HttpClient, and gRPC client instrumentation libraries to work without ActivitySourceAdapter.

## Changes
- Removed `ActivitySourceAdapter`
- Removed `AddDiagnosticSourceInstrumentation`
- Added a public API called `AddLegacyOperationName` which creates an activity listener for the empty ActivitySourceName and 
maintains a list of operation names to be looked up against on ActivityStarted callback to decide if the legacy activity should be processed
- Added the sampling check in ActivityListener OnStarted to sample the activities generated by instrumentations correctly
- Added the instrumentation ActivitySource such as "OpenTelemetry.Instrumentation.AspNetCore" and "OpenTelemetry.Instrumentation.Http" to `sources` in `AddInstrumentation` method of `TracerProviderBuilderSdk.cs`. ActivitySource is updated from an empty string to the assembly name in `OnStartActivity` of the instrumentation listeners, therefore we need to add these assembly names as a Source to the provider so that ActivityStopped callback can be called.
- Moved the code to set ActivitySource and Activity Kind for instrumented activities to `ActivityInstrumentationHelper.cs` under OpenTelemetry\Internal which is used by both AspNetCore and HttpClient instrumentation libraries.

* [x] `CHANGELOG.md` updated for non-trivial changes

### CODE REVIEW

File: src/OpenTelemetry/Trace/TracerProviderBuilderExtensions.cs
Comment (cijothomas) [inline src/OpenTelemetry/Trace/TracerProviderBuilderExtensions.cs:95]: `AddDiagnosticSourceInstrumentation` method can now be eliminated and simply use `AddInstrumentation`

File: src/OpenTelemetry/Trace/TracerProviderBuilderSdk.cs
Comment (cijothomas) [inline src/OpenTelemetry/Trace/TracerProviderBuilderSdk.cs:177]: `DiagnosticSourceInstrumentationFactory` can be totally eliminated now.

File: src/OpenTelemetry/Trace/TracerProviderSdk.cs
Comment (cijothomas) [inline src/OpenTelemetry/Trace/TracerProviderSdk.cs:78]: we can have a list of Activity names, for which we need to run samplers, and only run this for those opt-in activity names. Running this for *all* the legacy activities would affect hot path performance.
(even with list of opt-in names, we need to run benchmarks to see if there is significant regression)

Comment (cijothomas) [reply src/OpenTelemetry/Trace/TracerProviderSdk.cs:78]: Something like the below:

List<string> LegacyActivityNames; which is populated when instrumentation calls `AddLegacyActivity(string name)`.

if this list is empty, then bypass the whole section. This would be the case when someone is using pure ActivitySource based activities only.

File: src/OpenTelemetry.Instrumentation.Http/TracerProviderBuilderExtensions.cs
Comment (cijothomas) [inline src/OpenTelemetry.Instrumentation.Http/TracerProviderBuilderExtensions.cs:63]:
```suggestion
            builder.AddInstrumentation(() => new HttpClientInstrumentation(httpClientOptions));
```

Comment (cijothomas) [reply src/OpenTelemetry.Instrumentation.Http/TracerProviderBuilderExtensions.cs:63]: followed by something like builder.AddLegacyActivity("nameofactivity from httpclient")

**Review by cijothomas — CHANGES_REQUESTED** [review]
Thanks! The overall approach looks good. Need to optimize the perf in hot path, so some bencmark results before/after would need to be seen.

**Comment by cijothomas** [comment]
@alanwest @CodeBlanch please take a look. This would eliminate the need of ActivitySourceAdapater, by making the TracerProviderSdk itself being aware of legacy activities.

File: src/OpenTelemetry/Trace/TracerProviderBuilderExtensions.cs
Comment (utpilla) [reply src/OpenTelemetry/Trace/TracerProviderBuilderExtensions.cs]: Removed `AddDiagnosticSourceInstrumentation` method

File: src/OpenTelemetry/Trace/TracerProviderBuilderSdk.cs
Comment (utpilla) [reply src/OpenTelemetry/Trace/TracerProviderBuilderSdk.cs]: Removed `DiagnosticSourceInstrumentationFactory`

File: src/OpenTelemetry/Trace/TracerProviderSdk.cs
Comment (utpilla) [reply src/OpenTelemetry/Trace/TracerProviderSdk.cs]: I have added a HashSet<string> to store all the operation names for the DiagnosticSource instrumentations that were added.

File: src/OpenTelemetry.Instrumentation.Http/TracerProviderBuilderExtensions.cs
Comment (utpilla) [reply src/OpenTelemetry.Instrumentation.Http/TracerProviderBuilderExtensions.cs]: I have created two new methods `AddInstrumentationActivitySource` and `AddLegacyActivityOperationName` which should be used in the `TracerProviderBuilderExtensions.cs` of the instrumentation projects such as AspNetCore and HttpClient.

`AddInstrumentationActivitySource` ensures that the ActivityListener's ActivityStop callback is called as the ActivitySource is updated by the OnStartActivity of instrumentation listeners.

`AddLegacyActivityOperationName` is used to limit the processing of activities in the activity listener to only those that are required.

File: src/OpenTelemetry/Trace/TracerProviderBuilderSdk.cs
Comment (cijothomas) [inline src/OpenTelemetry/Trace/TracerProviderBuilderSdk.cs:137]: we dont need this, as there is already `AddSource` and this is adding anything new not already offered by `AddSource`

File: src/OpenTelemetry/Trace/TracerProviderSdk.cs
Comment (cijothomas) [inline src/OpenTelemetry/Trace/TracerProviderSdk.cs:167]: We dont want to affect the regular flow here by subscribing to empty sources. Do that conditionally, based of if there atleast one entry in the hashset/other datastructure keeping list of legacy activity names.

File: src/OpenTelemetry/Trace/TracerProviderBuilderSdk.cs
Comment (utpilla) [reply src/OpenTelemetry/Trace/TracerProviderBuilderSdk.cs]: Removed this.

File: src/OpenTelemetry/Trace/TracerProviderSdk.cs
Comment (utpilla) [reply src/OpenTelemetry/Trace/TracerProviderSdk.cs]: Added the condition to check if the HashSet is not empty.

File: src/OpenTelemetry/Trace/TracerProviderSdk.cs
Comment (cijothomas) [inline src/OpenTelemetry/Trace/TracerProviderSdk.cs:39]:
```suggestion
        private bool supportLegacyActivity;
```
```suggestion
        private bool needInstrumentations;
```

File: src/OpenTelemetry/Trace/TracerProviderSdk.cs
Comment (cijothomas) [inline src/OpenTelemetry/Trace/TracerProviderSdk.cs:74]: if the activity is created using `ActivitySource` and it *happens* to have same operation name as one of the legacyActivityOperationNames, then we'd be running sampler for them again here, which would be incorrect. We need to do this only for legacy activities - i.e the ones which has ActivitySource empty.

File: src/OpenTelemetry/Trace/TracerProviderSdk.cs
Comment (utpilla) [reply src/OpenTelemetry/Trace/TracerProviderSdk.cs]: Renamed the field to supportLegacyActivity.

File: src/OpenTelemetry/Trace/TracerProviderSdk.cs
Comment (utpilla) [reply src/OpenTelemetry/Trace/TracerProviderSdk.cs]: Yeah this is something I missed. There could be cases where ActivitySource creates an activity with the same operation name as the one created by DiagnosticSource. I have added a check for empty ActivitySource.

File: src/OpenTelemetry/Trace/TracerProviderSdk.cs
Comment (cijothomas) [inline src/OpenTelemetry/Trace/TracerProviderSdk.cs:74]: How about the following. 
```
if (this.supportLegacyActivity && activity.Source.Name.Equals(string.Empty))
                    {
                        // We have a legacy activity in hand now.
                        // and it's name matches the one user is interested in
                        if (legacyActivityOperationNames.Contains(activity.OperationName))
                        {
                            this.getRequestedDataAction(activity);
                        }
                        else
                        {
                            // legacy activity doesn't match user configured list.
                            // no need to proceeding further.
                            return;
                        }
                    }
```

Comment (cijothomas) [reply src/OpenTelemetry/Trace/TracerProviderSdk.cs:74]: similar in ActivityStopped.
```
if (this.supportLegacyActivity && activity.Source.Name.Equals(string.Empty))
                    {
                        // We have a legacy activity in hand now
                        // but not something user expressed interest in.
                        // Dont proceed any further.
                        if (!legacyActivityOperationNames.Contains(activity.OperationName))
                        {
                            return;
                        }
                    }
```

**Comment by CodeBlanch** [comment]
I like this approach! It kind of makes what I'm doing on #1761 impossible, but probably better to approach that through the runtime team anyway 😄

File: src/OpenTelemetry/Trace/TracerProviderSdk.cs
Comment (utpilla) [reply src/OpenTelemetry/Trace/TracerProviderSdk.cs]: Fixed this.

**Comment by utpilla** [comment]
With this new approach, we have introduced a few conditional checks to the TracerProviderSdk hot path for processing activities. I ran some benchmark tests and found that this has increased the mean time to run for the Tracer Benchmark test by up to ~20ns.

I feel this is acceptable considering the convenience it brings by getting rid of ActivitySourceAdapter entirely.

Here are the benchmarking results:

BenchmarkDotNet=v0.12.1, OS=Windows 10.0.19042
Intel Core i7-9700 CPU 3.00GHz, 1 CPU, 8 logical and 8 physical cores
.NET Core SDK=5.0.103
  [Host]     : .NET Core 3.1.12 (CoreCLR 4.700.21.6504, CoreFX 4.700.21.6905), X64 RyuJIT
  DefaultJob : .NET Core 3.1.12 (CoreCLR 4.700.21.6504, CoreFX 4.700.21.6905), X64 RyuJIT


**With the new changes:**

|                    Method |      Mean |    Error |   StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|-------------------------- |----------:|---------:|---------:|-------:|------:|------:|----------:|
|                NoListener |  23.50 ns | 0.034 ns | 0.058 ns |      - |     - |     - |         - |
|    PropagationDataListner | 688.19 ns | 2.647 ns | 2.476 ns | 0.1078 |     - |     - |     680 B |
|            AllDataListner | 684.48 ns | 3.639 ns | 3.404 ns | 0.1078 |     - |     - |     680 B |
| AllDataAndRecordedListner | 687.80 ns | 2.414 ns | 2.140 ns | 0.1078 |     - |     - |     680 B |
|              OneProcessor | 716.29 ns | 1.442 ns | 1.349 ns | 0.1078 |     - |     - |     680 B |
|             TwoProcessors | 736.54 ns | 5.225 ns | 4.632 ns | 0.1078 |     - |     - |     680 B |
|           ThreeProcessors | 743.95 ns | 3.218 ns | 2.687 ns | 0.1078 |     - |     - |     680 B |
|        OneInstrumentation | 730.72 ns | 1.876 ns | 1.663 ns | 0.1078 |     - |     - |     680 B |
|       TwoInstrumentations | 708.67 ns | 1.221 ns | 1.020 ns | 0.1078 |     - |     - |     680 B |

I added two Benchmark tests called `OneInstrumentation` and `TwoInstrumentations`:
- `OneInstrumentation` adds `AspNetCoreInstrumentation` to the provider
- `TwoInstrumentations` adds `AspNetCoreInstrumentation` and `HttpClientInstrumentation` to the provider
Both of these tests also add one processor to the TracerProvider.



**Without these changes (main branch):**

|                    Method |      Mean |    Error |   StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|-------------------------- |----------:|---------:|---------:|-------:|------:|------:|----------:|
|                NoListener |  23.45 ns | 0.034 ns | 0.042 ns |      - |     - |     - |         - |
|    PropagationDataListner | 677.92 ns | 1.638 ns | 1.368 ns | 0.1078 |     - |     - |     680 B |
|            AllDataListner | 685.51 ns | 2.517 ns | 2.354 ns | 0.1078 |     - |     - |     680 B |
| AllDataAndRecordedListner | 684.66 ns | 1.621 ns | 1.354 ns | 0.1078 |     - |     - |     680 B |
|              OneProcessor | 720.67 ns | 1.455 ns | 1.361 ns | 0.1078 |     - |     - |     680 B |
|             TwoProcessors | 711.47 ns | 2.183 ns | 1.823 ns | 0.1078 |     - |     - |     680 B |
|           ThreeProcessors | 721.47 ns | 1.324 ns | 1.239 ns | 0.1078 |     - |     - |     680 B |

File: src/OpenTelemetry.Instrumentation.AspNet/TracerProviderBuilderExtensions.cs
Comment (cijothomas) [inline src/OpenTelemetry.Instrumentation.AspNet/TracerProviderBuilderExtensions.cs:46]: Why is this needed?

Comment (utpilla) [reply src/OpenTelemetry.Instrumentation.AspNet/TracerProviderBuilderExtensions.cs:46]: We need to add the source as the instrumented activity's source will be updated from empty string to this in the OnStartActivityMethod. The ActivityStopped callback method checks if the activity's ActivitySource is present in the list of sources before calling Processor.OnEnd

Comment (cijothomas) [reply src/OpenTelemetry.Instrumentation.AspNet/TracerProviderBuilderExtensions.cs:46]: > The ActivityStopped callback method checks if the activity's ActivitySource is present in the list of sources before calling Processor.OnEnd

I dont see that check. ActivityStopped is simply checking if ActivitySource is empty or not.

Comment (utpilla) [reply src/OpenTelemetry.Instrumentation.AspNet/TracerProviderBuilderExtensions.cs:46]: Ahh yes! I should update the ActivityStopped callback to check if the ActivitySource was added to the provider. This would require us to have a private field for Sources which will have all the sources that were added to the provider using AddSource. Does that sound good?

Comment (utpilla) [reply src/OpenTelemetry.Instrumentation.AspNet/TracerProviderBuilderExtensions.cs:46]: The check for ActivitySource happens on `listener.ShouldListenTo`. If the source is not added then the ActivityStop callback will not be called `listener.ShouldListenTo` will return false for such an activity.

File: test/OpenTelemetry.Tests/Trace/TracerProviderSdkTest.cs
Comment (cijothomas) [inline test/OpenTelemetry.Tests/Trace/TracerProviderSdkTest.cs:310]: this is unused?

Comment (utpilla) [reply test/OpenTelemetry.Tests/Trace/TracerProviderSdkTest.cs:310]: There is an Assert statement for it in the beginning of `StartAction` in the `TestActivityProcessor`.

File: src/OpenTelemetry.Instrumentation.AspNet/TracerProviderBuilderExtensions.cs
Comment (cijothomas) [inline src/OpenTelemetry.Instrumentation.AspNet/TracerProviderBuilderExtensions.cs:46]: please move the sourcename to a constant. There is inconsistency now as this uses AspNetInstrumentation vs HttpInListener actually. (name is ultimately same, but lets do this consistently)

File: src/OpenTelemetry.Instrumentation.AspNetCore/AspNetCoreInstrumentation.cs
Comment (cijothomas) [inline src/OpenTelemetry.Instrumentation.AspNetCore/AspNetCoreInstrumentation.cs:33]: I initially thought (inccoreclty) this is a breaking change! The public api analyzer is a gift!

File: src/OpenTelemetry.Instrumentation.AspNetCore/AspNetCoreInstrumentation.cs
Comment (cijothomas) [inline src/OpenTelemetry.Instrumentation.AspNetCore/AspNetCoreInstrumentation.cs:35]: there are lot of special names - lets move them to a common place where it can be commented and probbaly linked to aspnetcore repo source code.
(separate PR)

File: src/OpenTelemetry.Instrumentation.AspNet/TracerProviderBuilderExtensions.cs
Comment (cijothomas) [inline src/OpenTelemetry.Instrumentation.AspNet/TracerProviderBuilderExtensions.cs:47]: dont we create sibling activities in asp.net also?

File: src/OpenTelemetry/Trace/TracerProviderBuilderExtensions.cs
Comment (cijothomas) [inline src/OpenTelemetry/Trace/TracerProviderBuilderExtensions.cs:81]:
```suggestion
        public static TracerProviderBuilder AddLegacyActivity(this TracerProviderBuilder tracerProviderBuilder, string operationName)
```

File: src/OpenTelemetry/Trace/TracerProviderBuilderExtensions.cs
Comment (cijothomas) [inline src/OpenTelemetry/Trace/TracerProviderBuilderExtensions.cs:75]: /// Adds activity with given operation name to the list of subscribed activities. This is for legacy activities (i.e activities created without using ActivituSource API).

File: src/OpenTelemetry/Trace/TracerProviderBuilderSdk.cs
Comment (cijothomas) [inline src/OpenTelemetry/Trace/TracerProviderBuilderSdk.cs:133]: its misleading to say "created by DiagnosticSource instrumentation".

File: test/Benchmarks/Trace/TraceBenchmarks.cs
Comment (cijothomas) [inline test/Benchmarks/Trace/TraceBenchmarks.cs:89]: This benchmark has nothing to do with `AddAspNetCoreInstrumentation`. You can simply use the raw API directly - AddLegacyActivity("somename")

**Review by cijothomas — APPROVED** [review]
LGTM.
Left a few comments, nothing blocking. Okay to have them as follow up PRs.

File: src/OpenTelemetry.Instrumentation.AspNet/TracerProviderBuilderExtensions.cs
Comment (utpilla) [reply src/OpenTelemetry.Instrumentation.AspNet/TracerProviderBuilderExtensions.cs]: I have updated the code to use ActivitySourceName from the listener implementation for all four instrumentation libraries

File: src/OpenTelemetry.Instrumentation.AspNetCore/AspNetCoreInstrumentation.cs
Comment (utpilla) [reply src/OpenTelemetry.Instrumentation.AspNetCore/AspNetCoreInstrumentation.cs:33]: Should we change this to internal?

File: src/OpenTelemetry.Instrumentation.AspNet/TracerProviderBuilderExtensions.cs
Comment (utpilla) [reply src/OpenTelemetry.Instrumentation.AspNet/TracerProviderBuilderExtensions.cs]: Thanks! I missed this one. I have added it.

File: src/OpenTelemetry/Trace/TracerProviderBuilderExtensions.cs
Comment (utpilla) [reply src/OpenTelemetry/Trace/TracerProviderBuilderExtensions.cs]: I have updated the name to AddLegacyActivity and updated the public API docs as well.

File: src/OpenTelemetry/Trace/TracerProviderBuilderExtensions.cs
Comment (utpilla) [reply src/OpenTelemetry/Trace/TracerProviderBuilderExtensions.cs]: I have updated the description.

File: src/OpenTelemetry/Trace/TracerProviderBuilderSdk.cs
Comment (utpilla) [reply src/OpenTelemetry/Trace/TracerProviderBuilderSdk.cs]: I have updated the description for this to match the public extension method's description.

File: test/Benchmarks/Trace/TraceBenchmarks.cs
Comment (utpilla) [reply test/Benchmarks/Trace/TraceBenchmarks.cs]: I have removed the instrumentations and used AddLegacyActivity.

File: src/OpenTelemetry.Instrumentation.GrpcNetClient/TracerProviderBuilderExtensions.cs
Comment (eddynaka) [inline src/OpenTelemetry.Instrumentation.GrpcNetClient/TracerProviderBuilderExtensions.cs:49]: nit: create as a const, just like aspnetcore

File: src/OpenTelemetry.Instrumentation.Http/TracerProviderBuilderExtensions.cs
Comment (eddynaka) [inline src/OpenTelemetry.Instrumentation.Http/TracerProviderBuilderExtensions.cs:63]: nit: create const

File: src/OpenTelemetry/Internal/ActivityInstrumentationHelper.cs
Comment (eddynaka) [inline src/OpenTelemetry/Internal/ActivityInstrumentationHelper.cs:23]: should we compile the existing extension instead of duplicating it?

File: src/OpenTelemetry/Trace/TracerProviderBuilderSdk.cs
Comment (eddynaka) [inline src/OpenTelemetry/Trace/TracerProviderBuilderSdk.cs:137]: nit: we could change this to `params string[]`, we have many cases calling this twice.

**Comment by cijothomas** [comment]
@open-telemetry/dotnet-approvers Please take a look. Will proceed to merge this in next few hours as 1st step in unblocking instrumentations in contrib repo. 
This is adding new public API, if there are feedbacks about the name, etc - we have opportunity to change it before the next stable release.

File: src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInListener.cs
Comment (CodeBlanch) [inline src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInListener.cs:103]: @cijothomas Are we exporting these in 1.0? Could technically be breaking if so? I think it's worth doing but should be called out as a change. Might already be in changelog I haven't gotten that far in the diff yet 😄

File: src/OpenTelemetry/.publicApi/net452/PublicAPI.Unshipped.txt
Comment (CodeBlanch) [inline src/OpenTelemetry/.publicApi/net452/PublicAPI.Unshipped.txt:1]: Not a big fan of this name. How about `AddLegacySource`? Given we have `AddSource` already.

File: src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInListener.cs
Comment (cijothomas) [reply src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInListener.cs:103]: Instrumentation haven't reached 1.0, so okay to make breaking change. However, I dont think there is any breaking change (breaking behaviour). Previously we were exporting only one activity, which is the one we create, if we create one. The framework is not exported today, if we create sibling activity ourselves

File: src/OpenTelemetry/.publicApi/net452/PublicAPI.Unshipped.txt
Comment (cijothomas) [reply src/OpenTelemetry/.publicApi/net452/PublicAPI.Unshipped.txt:1]: I am okay with `AddLegacySource`

Comment (cijothomas) [reply src/OpenTelemetry/.publicApi/net452/PublicAPI.Unshipped.txt:1]: we have opportunity to rename this until 1.1 stable is out, so i'd prefer to not block this pr.

Comment (CodeBlanch) [reply src/OpenTelemetry/.publicApi/net452/PublicAPI.Unshipped.txt:1]: Fine by me, I'll stamp as approved.

**Review by CodeBlanch — APPROVED** [review]

**Pull request merged**
